design-documents/service-isolation/inter-service-communication.md

```diff
@@ -0,0 +1,319 @@
+to _then_ callback and an exception to _otherwise_ callback if something during the remote call went wrong (like a network error). The
+service execution result may contain the return value of a service or an exception it has thrown. The promise that invoker returns may
+be used in both synchronous and asynchronous ways to be able to work with both types of services. More details about the promise
+[here](../promises.md).
+ 
+Proxy DTOs will look like this:
+```php
+namespace Magento\ServiceBProxy;
+
+class BDataProxy implements BDataInterface
+{
+    private $id;
+    
+    private $name;
+    
+    public function __construct(string $name, ?string $id) {
+        $this->name = $name;
+        $this->id = $id;
+    }
+    
+    public function getId(): ?string
+    {
+        return $this->id;
+    }
+    
+    public function setId(string $id): void
+    {
+        $this->id = $id;
+    }
+    
+    public function getName(): string
+    {
+        return $this->name;
+    }
+    
+    public function setName(string $name): void
+    {
+        $this->name = $name;
+    }
+}
+```
+Usually DTOs implementations are models in real modules but we need a class to be initialized when client code communicates with a
+remote service and for that simple DTOs will suffice.
+ 
+#### The invoker
+The invoker will be a part of Magento\Framework and will be present in every node installation. Invoker's job is to find a remote node's
+address based on given class name and method, send the request, receive response and convert it to either requested service method return type
+or an exception if it was thrown by the remote service. It returns a promise to support asynchronous services and because
+the requests to remote services will be asynchronous (depends on implementation). More details on the invoker's API and SPI
+[here](invoker.md).
+ 
+#### Gateway
+For actually delivering remote service requests RESTful API will be used - it's an established way to execute requested
+class' methods and to
+serialize arguments and return values. Existing web API gateway will be used as an RPC to save time on developing new
+gateway - there is no need to preserve RESTful principles (like using certain HTTP methods for certain types of operations)
+for inter-service communications since requests will be generated automatically and no human developer will be working
+with inter APIs explicitly.
+ 
+For sending HTTP requests [Guzzle](https://github.com/guzzle/guzzle) library will be used - it's
+a stable and well supported library which is capable of sending concurrent requests and allows wide customization.
+Guzzle will be configured to use cUrl (cUrl multi) as a backend. When asynchronous requests are sent via Guzzle it
+returns a promise which can easily be converted to Magento's promise.
+#### Lookup
+The information on how to reach services will be provided in env.php like this:
+```php
+return [
+    'remote_deployment' => [
```

Comment:
I propose something like this ['services' => ['catalog' => ['address' => 'catalog']]]. Whether we need to specify protocol is not clear at this point (depends on communication protocol we going to use), so I would suggest to omit it for now or make separate parameter.